pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -393,37 +407,43 @@ func (i *instantSplitter) splitAndSquashCall(expr *parser.Call, stats *MapperSta
 
 // assertSplittableRangeInterval returns the range interval specified in the input expr and whether it is greater than
 // the configured split interval.
-func (i *instantSplitter) assertSplittableRangeInterval(expr parser.Expr) (rangeInterval time.Duration, canSplit bool) {
-	rangeInterval = getRangeInterval(expr)
+func (i *instantSplitter) assertSplittableRangeInterval(expr parser.Expr) (rangeInterval time.Duration, canSplit bool, err error) {
+	rangeIntervals := getRangeIntervals(expr)
+	if len(rangeIntervals) == 0 {
+		return time.Duration(0), false, nil
+	}
+	if len(rangeIntervals) > 1 {
+		return time.Duration(0), false, fmt.Errorf("found %d range intervals while expecting at most 1", len(rangeIntervals))
+	}
+
+	rangeInterval = rangeIntervals[0]
 	if rangeInterval > i.interval {
-		return rangeInterval, true
+		return rangeInterval, true, nil
 	}
 
 	level.Debug(i.logger).Log("msg", "unable to split expression because range interval is smaller than configured split interval", "expr", expr, "range_interval", rangeInterval, "split_interval", i.interval)
-	return time.Duration(0), false
+	return time.Duration(0), false, nil
 }
 
-// getRangeInterval returns the range interval in the range vector expression
-// Returns 0 if no range interval is found
-// Example: expression `count_over_time({app="foo"}[10m])` returns 10m
-func getRangeInterval(expr parser.Expr) time.Duration {
-	switch e := expr.(type) {
-	case *parser.AggregateExpr:
-		return getRangeInterval(e.Expr)
-	case *parser.Call:
-		// Iterate over Call's arguments until a MatrixSelector and a valid range interval are found
-		for _, arg := range e.Args {
-			if argRangeInterval := getRangeInterval(arg); argRangeInterval != 0 {
-				return argRangeInterval
-			}
+// getRangeIntervals recursively visit the input expr and returns a slice containing all range intervals found.
+func getRangeIntervals(expr parser.Expr) []time.Duration {
+	// Due to how this function is used, we expect to always find at most 1 range interval
+	// so we preallocate it accordingly.
+	ranges := make([]time.Duration, 0, 1)
+
+	// Ignore the error since we never return it.
+	_, _ = anyNode(expr, func(entry parser.Node) (bool, error) {
+		switch e := entry.(type) {
+		case *parser.MatrixSelector:
+			ranges = append(ranges, e.Range)
+		case *parser.SubqueryExpr:
+			ranges = append(ranges, e.Range)
```

Comment:
This will work, and I understand that this function should be agnostic, but if this is called with a subquery (although we are currently checking that ahead) this will append a value. What do you think about appending a 0? 
Nit, as we also have tests to guarantee that subquery are not split. But I'm just afraid there could be a weird query that we are not thinking about 😅 
```suggestion
			ranges = append(ranges, time.Duration(0))
```

Reply:
If we return 0 we're just breaking the contract. Returning a 0 looks an hack to me.

Reply:
Okie, let's keep it this way 👍